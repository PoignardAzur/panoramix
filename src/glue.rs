//! Bunch of types mostly stolen from Crochet, to use as glue with Druid

#![allow(unused)]

use druid::Data;
use std::collections::HashMap;
use std::sync::Arc;

pub use druid::debug_state::DebugState;

pub struct GlobalEventCx<'a> {
    pub app_data: &'a mut DruidAppData,
}

impl<'a> GlobalEventCx<'a> {
    pub fn new(app_data: &'a mut DruidAppData) -> Self {
        Self { app_data }
    }
}

/// The type we use for app data for Druid integration.
///
/// Currently this is action queues.
///
/// It should probably be a vec of actions, but we can refine
/// later. For button clicks it doesn't matter.
#[derive(Clone, Data, Default)]
pub struct DruidAppData(Arc<HashMap<Id, Action>>);

/// Actions that can be produced by widgets,
#[derive(Clone)]
pub enum Action {
    Clicked,
    FutureResolved,
    TextChanged(String),
}

impl DruidAppData {
    pub(crate) fn queue_action(&mut self, id: Id, action: Action) {
        Arc::make_mut(&mut self.0).insert(id, action);
    }

    pub(crate) fn dequeue_action(&mut self, id: Id) -> Option<Action> {
        if self.0.contains_key(&id) {
            Arc::make_mut(&mut self.0).remove(&id)
        } else {
            None
        }
    }

    /// Report whether the id has a non-empty action queue.
    pub(crate) fn has_action(&self, id: Id) -> bool {
        self.0.contains_key(&id)
    }
}

// ---

use std::sync::atomic::{AtomicUsize, Ordering};

static ID_COUNTER: AtomicUsize = AtomicUsize::new(0);

/// An identifier for an element.
///
/// It's a bit heavy-handed to have this id as well as widget
/// id in Druid; likely the two concepts should be unified. But
#[derive(Clone, Copy, Debug, PartialOrd, PartialEq, Ord, Eq, Hash)]
pub struct Id(usize);

impl Id {
    /// Allocate a new unique id.
    pub fn new() -> Id {
        Id(ID_COUNTER.fetch_add(1, Ordering::Relaxed))
    }
}

// ---

use tracing_subscriber::prelude::*;
use tracing_subscriber::{fmt, EnvFilter};
use tracing_unwrap::ResultExt;

pub fn init_tracing() {
    let fmt_layer = fmt::layer().with_target(true);
    let filter_layer = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("warn"))
        .unwrap_or_log();

    tracing_subscriber::registry()
        .with(filter_layer)
        .with(fmt_layer)
        .init();
}
