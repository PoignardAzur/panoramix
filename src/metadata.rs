use derivative::Derivative;

/// Placeholder type for elements that don't raise events.
///
/// Equivalent to `!`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum NoEvent {}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
pub struct NoState;

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Copy(bound = ""),
    Default(bound = ""),
    PartialEq(bound = ""),
    Eq(bound = ""),
    Hash(bound = "")
)]
pub struct Metadata<ComponentEvent, ComponentState> {
    _marker: std::marker::PhantomData<(ComponentEvent, ComponentState)>,
}

impl<ComponentEvent, ComponentState> Metadata<ComponentEvent, ComponentState> {
    pub fn new() -> Self {
        Default::default()
    }
}
