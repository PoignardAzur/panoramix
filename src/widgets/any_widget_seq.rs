use crate::widget_sequence::FlexWidget;
use crate::widget_sequence::WidgetSequence;

use std::any::Any;

pub trait AnyWidgetSeq: Any {
    fn as_any(&self) -> &dyn Any;
    fn as_mut_any(&mut self) -> &mut dyn Any;
    fn any_widgets(&self) -> Vec<&dyn FlexWidget>;
    fn any_widgets_mut(&mut self) -> Vec<&mut dyn FlexWidget>;
}

impl<T> AnyWidgetSeq for T
where
    T: WidgetSequence + 'static,
{
    fn as_any(&self) -> &dyn Any {
        self
    }
    fn as_mut_any(&mut self) -> &mut dyn Any {
        self
    }

    fn any_widgets(&self) -> Vec<&dyn FlexWidget> {
        self.widgets()
    }

    fn any_widgets_mut(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.widgets_mut()
    }
}

// -

pub struct WidgetSeqBox {
    pub value: Box<dyn AnyWidgetSeq>,
}

impl WidgetSequence for WidgetSeqBox {
    fn widgets(&self) -> Vec<&dyn FlexWidget> {
        self.value.any_widgets()
    }

    fn widgets_mut(&mut self) -> Vec<&mut dyn FlexWidget> {
        self.value.any_widgets_mut()
    }
}
