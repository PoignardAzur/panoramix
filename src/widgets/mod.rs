//! Wrapper types around druid widgets.

pub mod any_widget_seq;
pub mod button_widget;
pub mod checkbox_widget;
pub mod empty_sequence;
pub mod flex_widget;
pub mod optional_widget;
pub mod single_widget;
pub mod textbox_widget;
pub mod widget_list;
pub mod widget_tuple;

pub use any_widget_seq::*;
pub use button_widget::*;
pub use checkbox_widget::*;
pub use empty_sequence::*;
pub use flex_widget::*;
pub use optional_widget::*;
pub use single_widget::*;
pub use textbox_widget::*;
pub use widget_list::*;
pub use widget_tuple::*;
