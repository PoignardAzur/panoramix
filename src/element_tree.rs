use crate::glue::{DruidAppData, GlobalEventCx};
use crate::widget_sequence::WidgetSequence;

use druid::{Env, EventCtx};
use std::any::Any;
use std::fmt::Debug;

/// Context type passed to all components when building them.
pub struct CompCtx {
    pub(crate) local_state: Box<dyn Any>,
}

impl CompCtx {
    /// Returns the local state of the current component instance.
    ///
    /// Panics if the generic type doesn't match the component's local state type.
    pub fn use_local_state<T: 'static>(&self) -> &T {
        self.local_state.downcast_ref::<T>().unwrap()
    }

    // TODO - add methods
    // use_lifecycle
    // get_vdom_context
}

/// Context required by [`VirtualDom::reconcile`]
pub struct ReconcileCtx<'a, 'b, 'c, 'd, 'e> {
    pub event_ctx: &'a mut EventCtx<'d, 'e>,
    pub data: &'b mut DruidAppData,
    pub env: &'c Env,
}

/// The trait implemented by all GUI elements.
///
/// Every type you use to explicitly create a GUI in Panoramix ([`Button`](crate::elements::Button), [`TextBox`](crate::elements::TextBox), any user-made component) implements Element. You usually don't need to worry about this trait unless you want to implement your own custom element.
///
/// For helper methods that can be called on all elements, see [`ElementExt`].
///
/// ## Note about template resolution
///
/// This trait is parameterized on two template types: `CpEvent` and `CpState`, which represent the event and local-state type of the parent component an element is built in. They are supposed to flow "inwards" with type inference, starting from the `-> impl Element<MyEvent, MyState>` return type of your function.
///
/// To give a concrete example:
///
/// ```rust
/// # use panoramix::{component, CompCtx, Column, Element, ElementExt};
/// # use panoramix::elements::{ButtonClick, Button, Label};
/// # type BuyItem = ButtonClick;
/// #
/// #[component]
/// fn StoreItem(ctx: &CompCtx, item_name: String) -> impl Element<BuyItem, u32> {
///     let item_count = ctx.use_local_state::<u32>();
///     Column!(
///         Label::new(format!("Item: {} x{}", item_name, item_count)),
///         Button::new("+")
///             .on_click(|item_count, _| {
///                 *item_count += 1;
///             }),
///         Button::new("Buy")
///             .bubble_up::<BuyItem>()
///     )
/// }
/// ```
///
/// In this example, because the return type is declared to be `-> impl Element<BuyItem, u32>`, all elements that are returned (Label, Button, Column) will be transitively inferred to implement `Element<BuyItem, u32>`.
///
/// The flip side of this is that constructing an element and not returning it (eg doing `let x = Button::new("...");` and then not using `x`) will lead to a compile error, because the compiler can't infer what `CpEvent` and `CpState` should be.
///
pub trait Element<CpEvent = NoEvent, CpState = ()>: Debug {
    /// The type of events this element can raise.
    ///
    /// This is the type that [`ElementExt::on`], [`ElementExt::map_event`] and [`ElementExt::bubble_up`] can take. It's different from the `CpEvent` generic parameter, which is the event the parent component emits.
    ///
    /// In the `StoreItem` example, the `Event` type of buttons is `ButtonClick`, and their `CpEvent` parameter is `BuyItem`.
    type Event;

    type AggregateChildrenState: Clone + Default + Debug + PartialEq;
    type BuildOutput: VirtualDom<
        CpEvent,
        CpState,
        Event = Self::Event,
        AggregateChildrenState = Self::AggregateChildrenState,
    >;

    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState);
}

// TODO - Include documentation about what a Virtual DOM is and where the name comes from.
pub trait VirtualDom<CpEvent, CpState>: Debug {
    type AggregateChildrenState: Clone + Default + Debug + PartialEq;
    type TargetWidgetSeq: WidgetSequence;

    type Event;

    // update_value is intended to enable memoize-style HOC
    // where instead of returning a vdom node, it returns
    // something along the lines of struct KeepEverythingAsItWas()
    // Ugh. I'm not explaining this well.
    fn update_value(&mut self, other: Self)
    where
        Self: Sized,
    {
        *self = other;
    }

    fn init_tree(&self) -> Self::TargetWidgetSeq;

    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    );

    // TODO - Rename methods
    #[allow(unused_variables)]
    fn process_event(
        &self,
        component_state: &mut CpState,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<CpEvent> {
        None
    }

    #[allow(unused_variables)]
    fn process_local_event(
        &self,
        component_state: &mut CpState,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        None
    }
}

/// Placeholder type for elements that don't raise events.
///
/// Equivalent to `!`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum NoEvent {}

// Used in unit tests
#[allow(dead_code)]
pub(crate) fn assign_empty_state_type(_elem: &impl Element<NoEvent, ()>) {}

#[allow(dead_code)]
pub(crate) fn assign_state_type<CpEvent, CpState, Elem: Element<CpEvent, CpState>>(_elem: &Elem) {}

use crate::elements::with_event::{ParentEvent, WithBubbleEvent, WithCallbackEvent, WithMapEvent};

/// Helper methods that can be called on all elements.
pub trait ElementExt<CpEvent, CpState>: Element<CpEvent, CpState> + Sized {
    fn on<EventParam, Cb: Fn(&mut CpState, EventParam)>(
        self,
        callback: Cb,
    ) -> WithCallbackEvent<CpEvent, CpState, EventParam, Self, Cb>
    where
        Self::Event: ParentEvent<EventParam>,
    {
        WithCallbackEvent {
            element: self,
            callback,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
        }
    }

    fn map_event<EventParam, EventReturn, Cb: Fn(&mut CpState, EventParam) -> Option<EventReturn>>(
        self,
        callback: Cb,
    ) -> WithMapEvent<CpEvent, CpState, EventParam, EventReturn, Self, Cb>
    where
        Self::Event: ParentEvent<EventParam>,
        CpEvent: ParentEvent<EventReturn>,
    {
        WithMapEvent {
            element: self,
            callback,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
            _comp_return: Default::default(),
        }
    }

    fn bubble_up<Event>(self) -> WithBubbleEvent<CpEvent, CpState, Event, Self>
    where
        Self::Event: ParentEvent<Event>,
        CpEvent: ParentEvent<Event>,
    {
        WithBubbleEvent {
            element: self,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
        }
    }
}

impl<CpEvent, CpState, ET: Element<CpEvent, CpState>> ElementExt<CpEvent, CpState> for ET {}
