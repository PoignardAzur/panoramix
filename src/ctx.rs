use crate::glue::DruidAppData;
use crate::metadata::{Metadata, NoState};
use druid::{Env, EventCtx};
use std::any::{type_name, Any, TypeId};

/// Context type passed to all components when building them.
pub struct CompCtx<'a> {
    // Yeah, we're using a cell, sorry. It's very local, though.
    pub(crate) called_use_metadata: std::cell::Cell<bool>,
    pub(crate) local_state: &'a dyn Any,
}

impl<'a> CompCtx<'a> {
    pub fn use_metadata<ComponentEvent: 'static, ComponentState: 'static>(
        &self,
    ) -> Metadata<ComponentEvent, ComponentState> {
        if self.called_use_metadata.get() {
            panic!("error: 'use_metadata' can only be called once per component")
        }
        self.called_use_metadata.set(true);
        Default::default()
    }

    /// Returns the local state of the current component instance.
    ///
    /// Panics if the generic type doesn't match the component's local state type.
    pub fn get_local_state<ComponentEvent: 'static, ComponentState: 'static>(
        &self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &'a ComponentState {
        #![allow(unused_variables)]
        if (*self.local_state).type_id() == TypeId::of::<NoState>() {
            panic!("error: 'get_local_state' cannot be called for a component whose root element isn't ComponentOutput")
        }
        self.local_state
            .downcast_ref::<ComponentState>()
            .expect(&format!(
            "internal type error: get_local_state expected {:?} ({}), parent component gave {:?}",
            TypeId::of::<ComponentState>(),
            type_name::<ComponentState>(),
            (*self.local_state).type_id(),
        ))
    }

    // TODO - add methods
    // use_lifecycle
    // get_vdom_context
}

/// Context required by [`VirtualDom::reconcile`]
pub struct ReconcileCtx<'a, 'b, 'c, 'd, 'e> {
    pub event_ctx: &'a mut EventCtx<'d, 'e>,
    pub data: &'b mut DruidAppData,
    pub env: &'c Env,
}

pub struct ProcessEventCtx<'e, 's> {
    pub event_queue: &'e mut dyn Any,
    pub state: &'s mut dyn Any,
}

impl<'e, 's> ProcessEventCtx<'e, 's> {
    pub fn event_queue<ComponentEvent: 'static, ComponentState: 'static>(
        &mut self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &mut Vec<ComponentEvent> {
        #![allow(unused_variables)]
        let type_id = (*self.event_queue).type_id();
        self.event_queue
            .downcast_mut::<Vec<ComponentEvent>>()
            .expect(&format!(
                "internal type error: event handler expected {:?} ({}), parent component gave {:?}",
                TypeId::of::<Vec<ComponentEvent>>(),
                type_name::<Vec<ComponentEvent>>(),
                type_id,
            ))
    }

    pub fn state<ComponentEvent: 'static, ComponentState: 'static>(
        &mut self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &mut ComponentState {
        #![allow(unused_variables)]
        let type_id = (*self.event_queue).type_id();
        self.state.downcast_mut::<ComponentState>().expect(&format!(
            "internal type error: event handler expected {:?} ({}), parent component gave {:?}",
            TypeId::of::<ComponentState>(),
            type_name::<ComponentState>(),
            type_id,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::metadata::NoEvent;
    use test_env_log::test;

    #[derive(Debug, PartialEq, Eq)]
    struct MyEvent(i32);

    #[test]
    fn event_queue() {
        let md: Metadata<MyEvent, NoState> = Default::default();

        let mut event_queue = Vec::<MyEvent>::new();
        let mut ctx = ProcessEventCtx {
            event_queue: &mut event_queue,
            state: &mut NoState,
        };
        ctx.event_queue(md).push(MyEvent(42));

        assert_eq!(event_queue, vec![MyEvent(42)],);
    }

    #[test]
    fn state() {
        let md: Metadata<NoEvent, i64> = Default::default();

        let mut state = 12345_i64;
        let mut ctx = ProcessEventCtx {
            event_queue: &mut Vec::<NoEvent>::new(),
            state: &mut state,
        };

        assert_eq!(*ctx.state(md), 12345_i64,);

        *ctx.state(md) = 123_i64;
        assert_eq!(state, 123_i64,);
    }
}
