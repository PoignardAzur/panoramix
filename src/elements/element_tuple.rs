#![allow(unused_attributes)]
#![allow(non_camel_case_types)]

use crate::ctx::{ProcessEventCtx, ReconcileCtx};
use crate::element_tree::{Element, VirtualDom};
use crate::elements::internals::EmptyElementData;
use crate::glue::GlobalEventCx;
use crate::metadata::{NoEvent, NoState};
use crate::widgets::WidgetTuple;

use tracing::instrument;

#[derive(Debug, Clone, Default, PartialEq, Eq, Hash)]
pub struct ElementTupleData<
    C0: VirtualDom,
    C1: VirtualDom,
    C2: VirtualDom,
    C3: VirtualDom,
    C4: VirtualDom,
    C5: VirtualDom,
    C6: VirtualDom,
    C7: VirtualDom,
    C8: VirtualDom,
    C9: VirtualDom,
    C10: VirtualDom,
    C11: VirtualDom,
>(
    pub C0,
    pub C1,
    pub C2,
    pub C3,
    pub C4,
    pub C5,
    pub C6,
    pub C7,
    pub C8,
    pub C9,
    pub C10,
    pub C11,
);

macro_rules! replace_ty {
    ($_t:tt >>> $sub:ty) => {
        $sub
    };
}

macro_rules! replace_expr {
    ($_t:tt >>> $sub:expr) => {
        $sub
    };
}

macro_rules! declare_stuff {
    ( $TupleName:ident ; $( $Type:ident ),* ; $( $Remainder:ident ),* ; $( $index:tt ),* ) => {

#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
pub struct $TupleName<
    $(
        $Type: Element,
    )*
>(
    $(
        pub $Type,
    )*
);

impl<
        $(
            $Type: Element,
        )*
    > Element for $TupleName<$($Type,)*>
{
    type Event = NoEvent;
    type ComponentState = NoState;
    type AggregateChildrenState = (
        $(
            $Type::AggregateChildrenState,
        )*
        $(replace_ty!(($Remainder) >>>
            ()
        ),)*
    );
    type BuildOutput = ElementTupleData<
        $(
            $Type::BuildOutput,
        )*
        $(replace_ty!(($Remainder) >>>
            EmptyElementData
        ),)*
    >;

    #[instrument(name = "Tuple", skip(self, prev_state))]
    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState) {
        let mut state : Self::AggregateChildrenState = Default::default();

        let node = ElementTupleData(
            $(
                {
                    let (subnode, substate) = self.$index.build(prev_state.$index);
                    state.$index = substate;
                    subnode
                },
            )*
            $(replace_expr!(($Remainder) >>>
                Default::default()
            ),)*
        );

        (node, state)
    }
}

    };
}

declare_stuff! {
    ElementTuple_1;
    T0; __, __, __, __, __, __, __, __, __, __, __ ;
    0
}

declare_stuff! {
    ElementTuple_2;
    T0, T1; __, __, __, __, __, __, __, __, __, __ ;
    0, 1
}

declare_stuff! {
    ElementTuple_3;
    T0, T1, T2; __, __, __, __, __, __, __, __, __ ;
    0, 1, 2
}

declare_stuff! {
    ElementTuple_4;
    T0, T1, T2, T3; __, __, __, __, __, __, __, __ ;
    0, 1, 2, 3
}

declare_stuff! {
    ElementTuple_5;
    T0, T1, T2, T3, T4; __, __, __, __, __, __, __ ;
    0, 1, 2, 3, 4
}

declare_stuff! {
    ElementTuple_6;
    T0, T1, T2, T3, T4, T5; __, __, __, __, __, __ ;
    0, 1, 2, 3, 4, 5
}

declare_stuff! {
    ElementTuple_7;
    T0, T1, T2, T3, T4, T5, T6; __, __, __, __, __ ;
    0, 1, 2, 3, 4, 5, 6
}

declare_stuff! {
    ElementTuple_8;
    T0, T1, T2, T3, T4, T5, T6, T7; __, __, __, __ ;
    0, 1, 2, 3, 4, 5, 6, 7
}

declare_stuff! {
    ElementTuple_9;
    T0, T1, T2, T3, T4, T5, T6, T7, T8; __, __, __ ;
    0, 1, 2, 3, 4, 5, 6, 7, 8
}

declare_stuff! {
    ElementTuple_10;
    T0, T1, T2, T3, T4, T5, T6, T7, T8, T9; __, __ ;
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9
}

declare_stuff! {
    ElementTuple_11;
    T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10; __ ;
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
}

declare_stuff! {
    ElementTuple_12;
    T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11 ;;
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11
}

/// Builds a group of up to 12 Elements, without a specified layout.
///
/// Return value implements [`Element`](crate::Element) as well.
///
/// ## Events
///
/// Returned element doesn't emit events.
#[macro_export]
macro_rules! Tuple {
    ( $(,)? ) => {
        $crate::elements::EmptyElement::new()
    };

    ( $e0:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_1($e0)
    };
    ( $e0:expr, $e1:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_2($e0, $e1)
    };
    ( $e0:expr, $e1:expr, $e2:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_3($e0, $e1, $e2)
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_4($e0, $e1, $e2, $e3)
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_5($e0, $e1, $e2, $e3, $e4)
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_6($e0, $e1, $e2, $e3, $e4, $e5)
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr, $e6:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_7($e0, $e1, $e2, $e3, $e4, $e5, $e6)
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr, $e6:expr, $e7:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_8($e0, $e1, $e2, $e3, $e4, $e5, $e6, $e7)
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr, $e6:expr, $e7:expr, $e8:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_9($e0, $e1, $e2, $e3, $e4, $e5, $e6, $e7, $e8)
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr, $e6:expr, $e7:expr, $e8:expr, $e9:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_10(
            $e0, $e1, $e2, $e3, $e4, $e5, $e6, $e7, $e8, $e9,
        )
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr, $e6:expr, $e7:expr, $e8:expr, $e9:expr, $e10:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_11(
            $e0, $e1, $e2, $e3, $e4, $e5, $e6, $e7, $e8, $e9, $e10,
        )
    };
    ( $e0:expr, $e1:expr, $e2:expr, $e3:expr, $e4:expr, $e5:expr, $e6:expr, $e7:expr, $e8:expr, $e9:expr, $e10:expr, $e11:expr $(,)? ) => {
        $crate::elements::element_tuple::ElementTuple_12(
            $e0, $e1, $e2, $e3, $e4, $e5, $e6, $e7, $e8, $e9, $e10, $e11,
        )
    };
}

impl<
        C0: VirtualDom,
        C1: VirtualDom,
        C2: VirtualDom,
        C3: VirtualDom,
        C4: VirtualDom,
        C5: VirtualDom,
        C6: VirtualDom,
        C7: VirtualDom,
        C8: VirtualDom,
        C9: VirtualDom,
        C10: VirtualDom,
        C11: VirtualDom,
    > VirtualDom for ElementTupleData<C0, C1, C2, C3, C4, C5, C6, C7, C8, C9, C10, C11>
{
    type Event = NoEvent;
    type AggregateChildrenState = (
        C0::AggregateChildrenState,
        C1::AggregateChildrenState,
        C2::AggregateChildrenState,
        C3::AggregateChildrenState,
        C4::AggregateChildrenState,
        C5::AggregateChildrenState,
        C6::AggregateChildrenState,
        C7::AggregateChildrenState,
        C8::AggregateChildrenState,
        C9::AggregateChildrenState,
        C10::AggregateChildrenState,
        C11::AggregateChildrenState,
    );
    type TargetWidgetSeq = WidgetTuple<
        C0::TargetWidgetSeq,
        C1::TargetWidgetSeq,
        C2::TargetWidgetSeq,
        C3::TargetWidgetSeq,
        C4::TargetWidgetSeq,
        C5::TargetWidgetSeq,
        C6::TargetWidgetSeq,
        C7::TargetWidgetSeq,
        C8::TargetWidgetSeq,
        C9::TargetWidgetSeq,
        C10::TargetWidgetSeq,
        C11::TargetWidgetSeq,
    >;

    #[instrument(name = "Tuple", skip(self))]
    fn init_tree(&self) -> Self::TargetWidgetSeq {
        WidgetTuple(
            self.0.init_tree(),
            self.1.init_tree(),
            self.2.init_tree(),
            self.3.init_tree(),
            self.4.init_tree(),
            self.5.init_tree(),
            self.6.init_tree(),
            self.7.init_tree(),
            self.8.init_tree(),
            self.9.init_tree(),
            self.10.init_tree(),
            self.11.init_tree(),
        )
    }

    #[instrument(name = "Tuple", skip(self, prev_value, widget_seq, ctx))]
    fn reconcile(
        &self,
        prev_value: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        self.0.reconcile(&prev_value.0, &mut widget_seq.0, ctx);
        self.1.reconcile(&prev_value.1, &mut widget_seq.1, ctx);
        self.2.reconcile(&prev_value.2, &mut widget_seq.2, ctx);
        self.3.reconcile(&prev_value.3, &mut widget_seq.3, ctx);
        self.4.reconcile(&prev_value.4, &mut widget_seq.4, ctx);
        self.5.reconcile(&prev_value.5, &mut widget_seq.5, ctx);
        self.6.reconcile(&prev_value.6, &mut widget_seq.6, ctx);
        self.7.reconcile(&prev_value.7, &mut widget_seq.7, ctx);
        self.8.reconcile(&prev_value.8, &mut widget_seq.8, ctx);
        self.9.reconcile(&prev_value.9, &mut widget_seq.9, ctx);
        self.10.reconcile(&prev_value.10, &mut widget_seq.10, ctx);
        self.11.reconcile(&prev_value.11, &mut widget_seq.11, ctx);
    }

    #[rustfmt::skip]
    #[instrument(name = "Tuple", skip(self, comp_ctx, children_state, widget_seq, cx))]
    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) {
        self.0.process_event(comp_ctx, &mut children_state.0, &mut widget_seq.0, cx);
        self.1.process_event(comp_ctx, &mut children_state.1, &mut widget_seq.1, cx);
        self.2.process_event(comp_ctx, &mut children_state.2, &mut widget_seq.2, cx);
        self.3.process_event(comp_ctx, &mut children_state.3, &mut widget_seq.3, cx);
        self.4.process_event(comp_ctx, &mut children_state.4, &mut widget_seq.4, cx);
        self.5.process_event(comp_ctx, &mut children_state.5, &mut widget_seq.5, cx);
        self.6.process_event(comp_ctx, &mut children_state.6, &mut widget_seq.6, cx);
        self.7.process_event(comp_ctx, &mut children_state.7, &mut widget_seq.7, cx);
        self.8.process_event(comp_ctx, &mut children_state.8, &mut widget_seq.8, cx);
        self.9.process_event(comp_ctx, &mut children_state.9, &mut widget_seq.9, cx);
        self.10.process_event(comp_ctx, &mut children_state.10, &mut widget_seq.10, cx);
        self.11.process_event(comp_ctx, &mut children_state.11, &mut widget_seq.11, cx);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::elements::label::Label;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn empty_tuple() {
        let tuple = Tuple!();
        let tuple_data = tuple.clone().build(Default::default());

        assert_debug_snapshot!(tuple);
        assert_debug_snapshot!(tuple_data);
    }

    #[test]
    fn new_tuple_single_item() {
        let tuple = Tuple!(Label::new("Hello"));
        let tuple_data = tuple.clone().build(Default::default());

        assert_debug_snapshot!(tuple);
        assert_debug_snapshot!(tuple_data);
    }

    #[test]
    fn new_tuple_multi_items() {
        let tuple = Tuple!(
            Label::new("Hello"),
            Label::new("Hello2"),
            Label::new("Hello3")
        );
        let tuple_trailing_comma = Tuple!(
            Label::new("Hello"),
            Label::new("Hello2"),
            Label::new("Hello3"),
        );
        let tuple_data = tuple.clone().build(Default::default());

        assert_debug_snapshot!(tuple);
        assert_debug_snapshot!(tuple_data);

        assert_eq!(tuple, tuple_trailing_comma);
    }

    #[test]
    fn label_tuple_widget() {
        use crate::test_harness::Harness;

        let tuple = Tuple!(
            Label::new("Hello"),
            Label::new("Hello2"),
            Label::new("Hello3"),
        );

        Harness::run_test_window(tuple, |harness| {
            let tuple_state = harness.get_root_debug_state();
            assert_debug_snapshot!(tuple_state);

            let new_tuple = Tuple!(
                Label::new("World"),
                Label::new("World2"),
                Label::new("World3"),
            );
            harness.update_root_element(new_tuple);

            let tuple_state_2 = harness.get_root_debug_state();
            assert_debug_snapshot!(tuple_state_2);
        });
    }
}
