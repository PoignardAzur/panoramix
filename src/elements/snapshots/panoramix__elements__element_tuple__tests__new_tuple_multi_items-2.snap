---
source: src/elements/element_tuple.rs
expression: tuple_data

---
(
    ElementTupleData(
        LabelData {
            text: "Hello",
            flex: FlexParams {
                flex: None,
                alignment: None,
            },
        },
        LabelData {
            text: "Hello2",
            flex: FlexParams {
                flex: None,
                alignment: None,
            },
        },
        LabelData {
            text: "Hello3",
            flex: FlexParams {
                flex: None,
                alignment: None,
            },
        },
        EmptyElementData,
        EmptyElementData,
        EmptyElementData,
        EmptyElementData,
        EmptyElementData,
        EmptyElementData,
        EmptyElementData,
        EmptyElementData,
        EmptyElementData,
    ),
    (
        (),
        (),
        (),
        (),
        (),
        (),
        (),
        (),
        (),
        (),
        (),
        (),
    ),
)
